### ISSUE TITLE
C<sort { return f() } LIST> fails to sort

### ISSUE THREAD

**Issue by ikegami** [Perl/perl5#18081]
```perl
#!/usr/bin/perl
use strict;
use warnings;
   
my @data = sort {
    return &special_sort;    # Same with: return special_sort();
} qw/ 6 1 2 /;
   
use Data::Dumper;
print STDERR Data::Dumper->Dump([\@data]);
   
sub special_sort {
    # (imagine some custom sorting logic here)
    return $a <=> $b;
}
```

Output:

```
Use of uninitialized value in sort at a.pl line 6.
Use of uninitialized value in sort at a.pl line 6.
Use of uninitialized value in sort at a.pl line 6.
$VAR1 = [
 6,
 1,
 2
];
````

My first thought is that `return` should be omitted. And indeed that works. But,

1. L<perlfunc/return> explicitly mentions that return can be used in a sort block.
2. return should throw an error if it's not allowed.

Originally "reported" (by someone else) on StackOverflow:
https://stackoverflow.com/q/63418268/589924

```
---
Flags:
    category=core
    severity=low
---
Site configuration information for perl 5.32.0:

Configured by ikegami at Sun Jun 28 18:05:50 EDT 2020.

Summary of my perl5 (revision 5 version 32 subversion 0) configuration:
   
  Platform:
    osname=linux
    osvers=4.4.0-18362-microsoft
    archname=x86_64-linux-thread-multi
    uname='linux tribble 4.4.0-18362-microsoft #836-microsoft mon may 05 16:04:00 pst 2020 x86_64 x86_64 x86_64 gnulinux '
    config_args='-de -Dprefix=/home/ikegami/usr/perlbrew/perls/5.32.0t -DPERL_SUB_DEPTH_WARN=1000 -Dusethreads -Aeval:scriptdir=/home/ikegami/usr/perlbrew/perls/5.32.0t/bin'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='cc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-O2'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='7.5.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/7/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.27.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.27'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'


---
@INC for perl 5.32.0:
    /home/ikegami/usr/perlbrew/perls/5.32.0t/lib/site_perl/5.32.0/x86_64-linux-thread-multi
    /home/ikegami/usr/perlbrew/perls/5.32.0t/lib/site_perl/5.32.0
    /home/ikegami/usr/perlbrew/perls/5.32.0t/lib/5.32.0/x86_64-linux-thread-multi
    /home/ikegami/usr/perlbrew/perls/5.32.0t/lib/5.32.0

---
Environment for perl 5.32.0:
    HOME=/home/ikegami
    LANG=C.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ikegami/usr/perlbrew/bin:/home/ikegami/usr/perlbrew/perls/latest/bin:.:/home/ikegami/bin:/home/ikegami/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/mnt/c/progs/sp5302-x64/c/bin:/mnt/c/progs/sp5302-x64/perl/site/bin:/mnt/c/progs/sp5302-x64/perl/bin:/mnt/c/Program Files/Intel/iCLS Client/:/mnt/c/Windows/system32:/mnt/c/Windows:/mnt/c/Windows/System32/Wbem:/mnt/c/Windows/System32/WindowsPowerShell/v1.0/:/mnt/c/Program Files (x86)/NVIDIA Corporation/PhysX/Common:/mnt/c/WINDOWS/system32:/mnt/c/WINDOWS:/mnt/c/WINDOWS/System32/Wbem:/mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0/:/mnt/c/WINDOWS/System32/OpenSSH/:/mnt/c/Program Files/IDM Computer Solutions/UltraEdit:/mnt/c/Program Files/PuTTY/:/mnt/c/Program Files/NVIDIA Corporation/NVIDIA NvDLISR:/mnt/c/Program Files (x86)/Intel/Intel(R) Management Engine
Components/DAL:/mnt/c/Program Files/Intel/Intel(R) Management Engine Components/DAL:/mnt/c/WINDOWS/system32:/mnt/c/WINDOWS:/mnt/c/WINDOWS/System32/Wbem:/mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0/:/mnt/c/WINDOWS/System32/OpenSSH/:/mnt/c/Users/ikegami/AppData/Roaming/ActiveState/bin:/mnt/c/Users/ikegami/AppData/Local/Microsoft/WindowsApps:/mnt/c/Users/ikegami/AppData/Local/atom/bin:/snap/bin
    PERLBREW_HOME=/home/ikegami/.perlbrew
    PERLBREW_MANPATH=/home/ikegami/usr/perlbrew/perls/latest/man
    PERLBREW_PATH=/home/ikegami/usr/perlbrew/bin:/home/ikegami/usr/perlbrew/perls/latest/bin
    PERLBREW_PERL=latest
    PERLBREW_ROOT=/home/ikegami/usr/perlbrew
    PERLBREW_SHELLRC_VERSION=0.88
    PERLBREW_VERSION=0.88
    PERL_BADLANG (unset)
    SHELL=/bin/bash
```

**Comment by tonycoz** [comment]
sort was added to this list in e5ad0abfbe2 by @iabyn , but it doesn't seem to have ever been consistently true.

Code like:
```
my $y = 0;
my @foo = sort { my $x; return $y+ss() } qw/5 2 1/;
```

does behave as expected, but the simple cases in this ticket don't, with similar behaviour as far back as 5.005_03.

**Comment by iabyn** [comment]
On Sun, Aug 23, 2020 at 10:24:16PM -0700, Tony Cook wrote:
> sort was added to this list in e5ad0abfbe2 by @iabyn , but it doesn't seem to have ever been consistently true.
> 
> Code like:
> ```
> my $y = 0;
> my @foo = sort { my $x; return $y+ss() } qw/5 2 1/;
> ```
> 
> does behave as expected, but the simple cases in this ticket don't, with
> similar behaviour as far back as 5.005_03.

(I've only just noticed there's a ticket for this issue - repeating my
posting of my p5p comment):

It's due to how the entersub op in the sort block is compiled. In

    sort {  special_sort() }

the entersub op (for calling special_sort) is marked as being in scalar
context (as shown by Concise):

    <1> entersub[t4] sKS/TARG ->(end)

while with

    sort { return special_sort() }

the entersub is marked as being in run-time-determined context:

    <1> entersub[t4] KS/TARG ->-

and then at run-time the corresponding leavesub op determines that the
context of the caller is void - hence the 'uninit value' warnings and
incorrect sort results when the return value is discarded.

I'm not sure yet whether it's the determining of the context at run-time
which is getting the wrong answer, or whether it's the marking the
entersub as run-time-determined which is wrong.

-- 
Never do today what you can put off till tomorrow.

**Comment by iabyn** [comment]
now fixed by v5.33.1-39-g390fe0c0d0
